package.json: Make llnode installable via npm

hhellyer opened the pull request:
This updates package.json and adds scripts to allow llnode to be
built via "npm install llnode"
It doesn't create a Node.js module you can use via require('llnode').
(This may change in the future depending on project direction.)

The npm name llnode has been reserved here:
https://www.npmjs.com/package/llnode
and Fedor also has access to that as discussed under issue #53.

This PR does leave the npm version number at 0.0.0, this should
be updated once the install process has been reviewed and at
that point the llnode package should be npm publish'd.

richardlau commented on package.json:
```diff
@@ -1,18 +1,31 @@
+  "//": "(Blame C++)",
   "scripts": {
+    "preinstall" : "node scripts/configure.js",
+    "install" : "./gyp_llnode && make -C out/",
```
I'm new to llnode, but is there a reason this couldn't be built with a `binding.gyp` which would automatically cause `npm` to invoke its bundled `node-gyp`?

hhellyer replied:
I was trying to leave the build process the same for cloning from git and installing via npm really.

rnchamberlain replied:
I wonder if we could support building the plugin in the 'clone from git' case using 'node-gyp rebuild' - that seems to be popular for node.js natives. Best ask @indutny

gibfahn replied:
>I was trying to leave the build process the same for cloning from git and installing via npm really.

Is there a reason you couldn't just `git clone https://github.com/nodejs/llnode.git && npm install` (i.e. use `node-gyp rebuild` for both)?

Fishrock123 replied:
^ That's what I was thinking?

hhellyer replied:
I think you'd have to ask @indutny why he originally didn't use node-gyp. I would add though that adding that to the build process adds a dependency on having node. When you build llnode from github there's not actually any dependancy on node.

There's a major dependancy on having the right lldb headers and you do need gyp in some form but weirdly for a node debugging tool there's no dependancy on node itself. llnode is built against a specific version of lldb but will work against a wide range of node versions - including being able to open Linux cores with lldb/llnode on Mac. If we did move to node-gyp we'd want to avoid introducing a link to the version of node that was running node-gyp.

For my part I wanted to focus on making the changes necessary to configure and build via npm to make it easier for people to get hold of rather than re-writing the build process. Moving to node-gyp isn't necessarily a bad idea but it should probably be raised as a separate issue.

indutny replied:
I'm fine with it, but let's handle it separately.

yjhjstz commented:
cool!

hhellyer commented:
I've updated configure.js to do a better job of finding the lldb executable on Linux.
It also now allows the user to force the use of a particular build of lldb by letting the user specify --lldb_exe on the command line. So on Ubuntu to force the user of lldb-3.9 you can do:

```npm install --lldb_exe=`which lldb-3.9` llnode```

hhellyer commented:
@indutny @Fishrock123 - I'll be out from the end of this week to 3rd January. I'm happy to make changes this week if you have any comments. If you don't have time this week I don't mind if you make changes on my branch while I'm out, otherwise we can fix it up in the new year.

hhellyer commented:
I've put this branch up on npmjs.org as version 0.0.1 so that it can get some testing and feedback when used as an actual npm.
https://www.npmjs.com/package/llnode

indutny commented on scripts/cleanup.js:
```diff
@@ -0,0 +1,19 @@
+var os = require('os');
```
Let's use ES6 😉 :

```
'use strict';
const os = ...
```

indutny commented on scripts/cleanup.js:
```diff
@@ -0,0 +1,19 @@
+
+var cwd = process.cwd();
+var osName = os.type();
+var libExt = "so";
```
`'` instead of `"`

indutny commented on scripts/cleanup.js:
```diff
@@ -0,0 +1,19 @@
+var llnodeLib = "llnode." + libExt;
+
+// Move the library somewhere easy to remember.
+console.log('Copying ' + cwd + '/out/Release/' + llnodeLib + ' to ' + cwd + '/' + llnodeLib);
```
Template strings:

```js
`Copying ${cwd} ...`
```

indutny commented on scripts/configure.js:
```diff
@@ -0,0 +1,192 @@
+  lldbHeadersBranch = lldbReleases[lldbVersion];
+  lldbIncludeDir = 'lldb-' + lldbVersion;
+
+} else if ( osName === 'Linux') {
```
Style: `if (osName === 'Linux') {`

indutny commented on scripts/configure.js:
```diff
@@ -0,0 +1,192 @@
+
+  lldbVersion = getDarwinRelease();
+
+  if(lldbVersion === undefined) {
```
`if (`

indutny replied:
I think we may want to use `eslint` here.

indutny commented on scripts/configure.js:
```diff
@@ -0,0 +1,192 @@
+function getLldbExecutable() {
+  var lldbExe = process.env.npm_config_lldb_exe;
+  if (lldbExe === undefined) {
+    var lldbExeNames = ['lldb', 'lldb-4.0', 'lldb-3.9', 'lldb-3.8', 'lldb-3.7', 'lldb-3.6']
```
Missing semicolon

rnchamberlain commented:
FYI, tried this out on Ubuntu, lldb 3.9, with a core dump from Node.js v7:
```
> npm install llnode
> lldb node -c core
(lldb) plugin load node_modules/llnode/llnode.so
(lldb) jsstack
```
Works well, this is going to be nice

hhellyer commented:
I copied the .eslintrc from a node build and ran that against the scripts. They should be formatted correctly now.

hhellyer commented:
@indutny I'll put the package.json level back to 1.4.0, do a merge and then publish a v1.4.0 package to npm. Is that reasonable? (We might need to document the publish process the contribution docs now too.)

indutny commented:
@hhellyer sounds reasonable to me.

hhellyer commented:
I've published to https://www.npmjs.com/package/llnode - it took an hour or so to update the npmjs.org page last time so that still shows 0.0.1 but doing `npm install llnode` now installs version 1.4.1.

indutny commented:
Fantastic! Thank you!

bnoordhuis commented:
@hhellyer Why did you rename/symlink llnode.gyp to llnode.gyp.json?  From looking at 2325ae3 it's quite clearly a deliberate move but for the life of me, I can't figure out why.

hhellyer commented:
@bnoordhuis - Well I'm quite prepared to be told I was being an idiot but no matter what I did I couldn't figure out why when I did npm install <path_to_folder> (during development) the npm never copied files ending in .gyp. So I renamed it llnode.gyp.json. (I've just re-tested it and calling it llnode.gyp or build.gyp both still cause it not to be copied.)

It might be an obscure npm feature or just something really silly I was doing but in the end renaming it seemed enough to solve my problem. There might be some weirdness around having .gyp files other than binding.gyp in your npm, I just added one to something else to verify the copying problem and it's existence alone broke the build.

I think I just put the symlink in so I didn't break anyone else who was referring to the file. (I was trying to make the minimum set of changes.)

richardlau commented:
I don't have the reference to hand but I believe npm has (had?) the surprising behaviour of renaming all `*.gyp` files to `binding.gyp`.

hhellyer commented:
@richardlau - That would actually fit what I saw. (And explain why now if I change binding.gyp in the npm I'm working on to iamafish.gyp it still builds!)

richardlau commented:
https://github.com/npm/npm/issues/8411 and https://github.com/npm/fstream-npm/pull/18
